## [INSERT TITLE]

- 
-
-
- Velasquez, Almira Zabrina Alyson Q.

link to dataset: https://www.start.umd.edu/data-tools/GTD?fbclid=IwY2xjawGsWdtleHRuA2FlbQIxMQABHelOWZ92imSyR5CNNl3hX7DyDm0RbLu7CRxasL54IcugUstkkzgQrBsLIg_aem_rWPWYEYQ7IHFXqqyMVw1ug

### Dataset Description

This dataset contains records on global terrorism incidents, including details such as attack types, locations, targets, and outcomes. It aims to provide comprehensive data for analyzing patterns and trends in terrorism across various regions and over time. The dataset likely includes key attributes such as:

- Date and location of incidents
- Attack methods (e.g., bombing, armed assault)
- Targets (e.g., civilians, infrastructure, government institutions)
- Outcomes (e.g., fatalities, injuries, damage)

##### The objective of this analysis is to:

- Identify global trends in terrorism by examining temporal and geographic patterns.

- Analyze common targets and their characteristics to understand which groups or entities are most frequently attacked.

- Determine correlations between types of attacks and their outcomes, such as the severity of casualties or economic impact.


In [ ]:
#IMPORTS

import os
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import LabelEncoder, StandardScaler
from geopy.distance import geodesic
from sklearn.feature_selection import mutual_info_classif, VarianceThreshold
from sklearn.preprocessing import LabelEncoder



In [100]:
file_path = 'globalterrorismdb_2021Jan-June_1222dist.xlsx'

df = pd.read_excel(file_path)

#### Exploratory Analysis

Describe your data and show what kind of initial features you are dealing with
You can point anomalies/outliers in the data. Discuss patterns or findings.

In [101]:
df.head()

eventid  iyear  imonth  iday  approxdate  extended resolution  \
0  202101010004   2021       1     1         NaN         0        NaT   
1  202101010005   2021       1     1  01/01/2021         0        NaT   
2  202101010006   2021       1     1         NaN         0        NaT   
3  202101010009   2021       1     1         NaN         0        NaT   
4  202101010024   2021       1     2         NaN         0        NaT   

   country  country_txt  region  ...  \
0       60        Egypt      10  ...   
1       92        India       6  ...   
2      228        Yemen      10  ...   
3        4  Afghanistan       6  ...   
4      182      Somalia      11  ...   

                                            addnotes  \
0                                                NaN   
1                                                NaN   
2  There is doubt that this incident meets terror...   
3  There is doubt that this incident meets terror...   
4                                                NaN   

                                              scite1  \
0  "Egyptian officials: Roadside bombing in Sinai...   
1  "Terror module busted, seven arrested in J&K,"...   
2  "Civilian reportedly dies of injuries from Hou...   
3  "Afghanistan- Ghor Provincial Council Member K...   
4  "Jubbaland lawmaker killed in car bomb blast,"...   

                                              scite2  \
0                                                NaN   
1  "Militants throw grenade on CRPF," HT Syndicat...   
2  "UN troubled by civilian deaths in Yemen's Hud...   
3  "Journalist killed on 1st day of new year in A...   
4  "Somali lawmaker killed in a sticky bomb attac...   

                                              scite3  \
0                                                NaN   
1                                                NaN   
2  "Five women killed in Yemen wedding attack," G...   
3  "Afghanistan: 2021 Begins with Assassination o...   
4  "Al-Shabab claims killing regional MP in south...   

                   dbsource  INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0  START Primary Collection       -9        -9        0      -9      NaN  
1  START Primary Collection       -9        -9        0      -9      NaN  
2  START Primary Collection        0         0        0       0      NaN  
3  START Primary Collection       -9        -9        0      -9      NaN  
4  START Primary Collection        0         0        0       0      NaN  

[5 rows x 135 columns]

In [102]:
df.tail()

eventid  iyear  imonth  iday  approxdate  extended resolution  \
4955  202106300023   2021       6    30         NaN         0        NaT   
4956  202106300029   2021       6    30  06/30/2021         0        NaT   
4957  202106300030   2021       6    30         NaN         1 2021-07-08   
4958  202106300038   2021       6    30         NaN         0        NaT   
4959  202107020037   2021       6    30         NaN         0        NaT   

      country  country_txt  region  ...  \
4955        4  Afghanistan       6  ...   
4956      138      Myanmar       5  ...   
4957      147      Nigeria      11  ...   
4958       45     Colombia       3  ...   
4959      138      Myanmar       5  ...   

                                               addnotes  \
4955                                                NaN   
4956                                                NaN   
4957                                                NaN   
4958                                                NaN   
4959  There is doubt that this incident meets terror...   

                                                 scite1  \
4955  "Gunmen blow up power pylon in Parwan," Afghan...   
4956  "Spring Revolution Daily News for 16-30 June 2...   
4957  "Boko Haram Releases Abducted Catholic Priest ...   
4958  "Two dead and one wounded after clashes betwee...   
4959  "Spring Revolution Daily News for 2-7 July 202...   

                                                 scite2  \
4955  "Provinces hit by blackout after power pylon d...   
4956                                                NaN   
4957  "Kidnapped Maiduguri Catholic Priest regains f...   
4958                                                NaN   
4959                                                NaN   

                                                 scite3  \
4955                                                NaN   
4956                                                NaN   
4957  "ISWAP-Boko Haram Abduct Catholic Priest In Bo...   
4958                                                NaN   
4959                                                NaN   

                      dbsource  INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
4955  START Primary Collection       -9        -9        0      -9      NaN  
4956  START Primary Collection       -9        -9        0      -9      NaN  
4957  START Primary Collection        0         0        0       0      NaN  
4958  START Primary Collection        0         0        0       0      NaN  
4959  START Primary Collection       -9        -9        0      -9      NaN  

[5 rows x 135 columns]

In [103]:
print("\nDataset Shape (Rows, Columns):", df.shape)


Dataset Shape (Rows, Columns): (4960, 135)


In [104]:
print("\nColumn Names:")
print(df.columns.tolist())



Column Names:
['eventid', 'iyear', 'imonth', 'iday', 'approxdate', 'extended', 'resolution', 'country', 'country_txt', 'region', 'region_txt', 'provstate', 'city', 'latitude', 'longitude', 'specificity', 'vicinity', 'location', 'summary', 'crit1', 'crit2', 'crit3', 'doubtterr', 'alternative', 'alternative_txt', 'multiple', 'success', 'suicide', 'attacktype1', 'attacktype1_txt', 'attacktype2', 'attacktype2_txt', 'attacktype3', 'attacktype3_txt', 'targtype1', 'targtype1_txt', 'targsubtype1', 'targsubtype1_txt', 'corp1', 'target1', 'natlty1', 'natlty1_txt', 'targtype2', 'targtype2_txt', 'targsubtype2', 'targsubtype2_txt', 'corp2', 'target2', 'natlty2', 'natlty2_txt', 'targtype3', 'targtype3_txt', 'targsubtype3', 'targsubtype3_txt', 'corp3', 'target3', 'natlty3', 'natlty3_txt', 'gname', 'gsubname', 'gname2', 'gsubname2', 'gname3', 'gsubname3', 'motive', 'guncertain1', 'guncertain2', 'guncertain3', 'individual', 'nperps', 'nperpcap', 'claimed', 'claimmode', 'claimmode_txt', 'claim2', 'clai

In [105]:
print("\nDataset Info:")
df.info()


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4960 entries, 0 to 4959
Columns: 135 entries, eventid to related
dtypes: datetime64[ns](1), float64(51), int64(29), object(54)
memory usage: 5.1+ MB


Rows: 4,960
Columns: 135
Data Types:
1 datetime column
51 float columns
29 integer columns
54 object columns (likely categorical or text)

#### Data Preprocessing and Cleaning


Explain why the data was preprocessed that way
If you removed data, explain why removing the data was necessary

##### Handle Missing Data

We check for missing values in each column computing the percentage.

In [106]:
missing_values = df.isnull().sum().sort_values(ascending=False)
missing_percentage = (missing_values / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage (%)': missing_percentage
}).reset_index().rename(columns={'index': 'Column'})

missing_df

Column  Missing Values  Percentage (%)
0    weapsubtype4_txt            4960           100.0
1       weaptype4_txt            4960           100.0
2           weaptype4            4960           100.0
3        weapsubtype4            4960           100.0
4           gsubname3            4960           100.0
..                ...             ...             ...
130          dbsource               0             0.0
131           INT_LOG               0             0.0
132          INT_IDEO               0             0.0
133          INT_MISC               0             0.0
134           INT_ANY               0             0.0

[135 rows x 3 columns]

Several columns have 100% missing values, indicating they might not be useful for analysis, we will proceed by removing these columns.
Some columns have partial missing data, we will procede by further analyzing the columns and decide which to impute or exclude in our analysis.

In [ ]:
columns_to_drop = missing_df[missing_df['Percentage (%)'] == 100]['Column'].tolist()
df_cleaned = df.drop(columns=columns_to_drop)

num_columns_dropped = len(columns_to_drop)

remaining_missing_values = df_cleaned.isnull().sum().sort_values(ascending=False)
remaining_missing_percentage = (remaining_missing_values / len(df_cleaned)) * 100

In [ ]:
remaining_missing_df = pd.DataFrame({
    'Column': remaining_missing_values.index,
    'Missing Values': remaining_missing_values.values,
    'Percentage (%)': remaining_missing_percentage.values
})


print("Number of Columns Dropped:", num_columns_dropped)
print("\nRemaining Missing Values Analysis:\n")
print(remaining_missing_df)

Number of Columns Dropped: 5

Remaining Missing Values Analysis:
             Column  Missing Values  Percentage (%)
0            divert            4959       99.979839
1     kidhijcountry            4959       99.979839
2        claimmode3            4959       99.979839
3    claimmode3_txt            4959       99.979839
4    claimmode2_txt            4953       99.858871
..              ...             ...             ...
125        dbsource               0        0.000000
126         INT_LOG               0        0.000000
127        INT_IDEO               0        0.000000
128        INT_MISC               0        0.000000
129         INT_ANY               0        0.000000

[130 rows x 3 columns]


In [ ]:
print("\nCleaned DataFrame Shape:", df_cleaned.shape)


Cleaned DataFrame Shape: (4960, 130)


We removed 5 columns with 100% missing values, reducing the dataset to 130 columns.

We will separate the numerical and categorical columns. For the numerical columns, we will fill the missing values with their median, and for the categorical columns, we will fill the missing values with 'Unknown'.

In [ ]:
numerical_cols = df_cleaned.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = df_cleaned.select_dtypes(include=['object']).columns

df_cleaned[numerical_cols] = df_cleaned[numerical_cols].fillna(df_cleaned[numerical_cols].median())

df_cleaned[categorical_cols] = df_cleaned[categorical_cols].fillna('Unknown')


remaining_missing_values_post = df_cleaned.isnull().sum().sum()
remaining_missing_values_post

np.int64(4864)

4864 still null. we identify columns with remaining missing values.

In [ ]:

remaining_missing_cols = df_cleaned.columns[df_cleaned.isnull().sum() > 0]

remaining_missing_analysis = pd.DataFrame({
    'Column': remaining_missing_cols,
    'Missing Values': df_cleaned[remaining_missing_cols].isnull().sum().values,
    'Percentage (%)': (df_cleaned[remaining_missing_cols].isnull().sum().values / len(df_cleaned)) * 100
})

print("Remaining Missing Columns Details:")
print(remaining_missing_analysis)


Remaining Missing Columns Details:
       Column  Missing Values  Percentage (%)
0  resolution            4864       98.064516


'resolution' column has missing values so we impute the missing values with 'unknown'.

In [ ]:
df_cleaned['resolution'] = df_cleaned['resolution'].fillna('Unknown')

final_missing_values = df_cleaned.isnull().sum().sum()

final_missing_values

np.int64(0)

We will analyze and count all categorical columns for unique values.

In [ ]:

categorical_cols = df_cleaned.select_dtypes(include='object').columns

categorical_analysis = pd.DataFrame({
    'Column': categorical_cols,
    'Unique Values': [df_cleaned[col].nunique() for col in categorical_cols],
    'Sample Values': [list(df_cleaned[col].unique()[:5]) for col in categorical_cols]  # Show a sample of 5 unique values
})

print("Categorical Columns Analysis:")
print(categorical_analysis)


Categorical Columns Analysis:
                Column  Unique Values  \
0           approxdate            165   
1           resolution             78   
2          country_txt             80   
3           region_txt             11   
4            provstate            443   
5                 city           2791   
6             location           1615   
7              summary           4881   
8      alternative_txt              6   
9      attacktype1_txt              9   
10     attacktype2_txt              6   
11     attacktype3_txt              3   
12       targtype1_txt             21   
13    targsubtype1_txt            100   
14               corp1           1470   
15             target1           1678   
16         natlty1_txt             88   
17       targtype2_txt             17   
18    targsubtype2_txt             66   
19               corp2            152   
20             target2            222   
21         natlty2_txt             53   
22       targtype3_txt     

In [ ]:
# Mapping transformations
region_mapping = {
    'north america': 'Developed Regions',
    'western europe': 'Developed Regions',
    'australasia & oceania': 'Developed Regions',
    'sub-saharan africa': 'Developing Regions',
    'south asia': 'Developing Regions',
    'middle east & north africa': 'Developing Regions',
    'southeast asia': 'Developing Regions',
    'central america & caribbean': 'Developing Regions',
    'south america': 'Developing Regions',
    'east asia': 'Developed Regions',
    'eastern europe': 'Developed Regions'
}
attack_type_mapping = {
    'hostage taking (kidnapping)': 'Hostage Situation',
    'hostage taking (barricade incident)': 'Hostage Situation',
    'bombing/explosion': 'Bombing',
    'armed assault': 'Armed Attack',
    'unarmed assault': 'Unarmed Attack',
    'facility/infrastructure attack': 'Infrastructure Attack'
}
target_type_mapping = {
    'private citizens & property': 'Civilians',
    'business': 'Civilians',
    'military': 'Military/Police',
    'police': 'Military/Police',
    'government (general)': 'Government',
    'government (diplomatic)': 'Government',
    'journalists & media': 'Media/NGOs',
    'ngo': 'Media/NGOs'
}

 We perform the data transformations and concatenate the results in the new data frame

In [ ]:

new_columns = pd.DataFrame({
    'region_txt_mapped': df_cleaned['region_txt'].map(region_mapping),
    'attacktype1_mapped': df_cleaned['attacktype1_txt'].map(attack_type_mapping).fillna(df_cleaned['attacktype1_txt']),
    'targtype1_mapped': df_cleaned['targtype1_txt'].map(target_type_mapping).fillna(df_cleaned['targtype1_txt'])
})

df_cleaned = pd.concat([df_cleaned, new_columns], axis=1)

print("Cleaned DataFrame Shape:", df_cleaned.shape)

transformed_preview = df_cleaned[['region_txt', 'region_txt_mapped', 'attacktype1_txt', 'attacktype1_mapped', 'targtype1_txt', 'targtype1_mapped']].head()
print("\nTransformed Categorical Mappings Preview:")
print(transformed_preview)


Cleaned DataFrame Shape: (4960, 133)

Transformed Categorical Mappings Preview:
                   region_txt region_txt_mapped    attacktype1_txt  \
0  Middle East & North Africa               NaN  Bombing/Explosion   
1                  South Asia               NaN  Bombing/Explosion   
2  Middle East & North Africa               NaN  Bombing/Explosion   
3                  South Asia               NaN      Assassination   
4          Sub-Saharan Africa               NaN      Assassination   

  attacktype1_mapped                targtype1_txt             targtype1_mapped  
0  Bombing/Explosion                     Military                     Military  
1  Bombing/Explosion                       Police                       Police  
2  Bombing/Explosion  Private Citizens & Property  Private Citizens & Property  
3      Assassination          Journalists & Media          Journalists & Media  
4      Assassination         Government (General)         Government (General)  


#### Feature Extraction

Explain (even if briefly) what these features are, and why they may help

In [ ]:
# insert feature extractionnn


Numerical Feature Importance:
         Feature  Mutual_Info_Score
22     targtype1           2.072833
23  targsubtype1           2.024043
13         crit3           0.438481
8      longitude           0.378439
14     doubtterr           0.343866
..           ...                ...
43     compclaim           0.000000
66   ransomamtus           0.000000
64        ransom           0.000000
61    nhostkidus           0.000000
67    ransompaid           0.000000

[75 rows x 2 columns]

Final Dataset with Selected Features:
   INT_ANY  longitude       eventid  region  doubtterr   latitude  crit3  \
0       -9  33.008587  202101010004      10          1  31.016779      0   
1       -9  74.917261  202101010005       6          0  34.010230      1   
2        0  42.973093  202101010006      10          1  14.782391      1   
3       -9  65.249242  202101010009       6          1  34.521315      1   
4        0  42.546652  202101010024      11          0  -0.357847      1   

  attacktype1_mapp

In [116]:
print("\nColumn Names:")
print(df_final.columns.tolist())


Column Names:
['INT_ANY', 'longitude', 'eventid', 'region', 'doubtterr', 'latitude', 'crit3', 'attacktype1_mapped', 'success', 'targtype1', 'region_txt_mapped', 'country', 'propextent', 'targtype1_mapped', 'natlty1', 'property', 'INT_MISC', 'targsubtype1', 'weaptype1', 'specificity', 'attacktype1', 'nkill', 'weapsubtype1']


In [117]:
df_final.head()

INT_ANY  longitude       eventid  region  doubtterr   latitude  crit3  \
0       -9  33.008587  202101010004      10          1  31.016779      0   
1       -9  74.917261  202101010005       6          0  34.010230      1   
2        0  42.973093  202101010006      10          1  14.782391      1   
3       -9  65.249242  202101010009       6          1  34.521315      1   
4        0  42.546652  202101010024      11          0  -0.357847      1   

  attacktype1_mapped  success  targtype1  ...             targtype1_mapped  \
0  Bombing/Explosion        1          4  ...                     Military   
1  Bombing/Explosion        1          3  ...                       Police   
2  Bombing/Explosion        1         14  ...  Private Citizens & Property   
3      Assassination        1         10  ...          Journalists & Media   
4      Assassination        1          2  ...         Government (General)   

   natlty1  property INT_MISC  targsubtype1  weaptype1  specificity  \
0     60.0         1        0          35.0          6            1   
1     92.0         0        0          22.0          6            1   
2    228.0         1        0          78.0          6            1   
3      4.0         1        0          54.0          5            1   
4    182.0         1        0          15.0          6            1   

   attacktype1  nkill  weapsubtype1  
0            3    2.0          12.0  
1            3    0.0           7.0  
2            3    5.0          11.0  
3            1    1.0           5.0  
4            1    1.0          29.0  

[5 rows x 23 columns]

In [118]:
df_final.tail()

INT_ANY  longitude       eventid  region  doubtterr   latitude  crit3  \
4955       -9  69.196838  202106300023       6          0  35.054772      1   
4956       -9  97.897143  202106300029       5          0  23.986739      1   
4957        0  13.014035  202106300030      11          0  11.572869      1   
4958        0 -76.333069  202106300038       3          0   3.104189      1   
4959       -9  97.895734  202107020037       5          1  23.974583      1   

               attacktype1_mapped  success  targtype1  ...  \
4955            Bombing/Explosion        1         21  ...   
4956  Hostage Taking (Kidnapping)        1          8  ...   
4957  Hostage Taking (Kidnapping)        1         15  ...   
4958            Bombing/Explosion        0         20  ...   
4959            Bombing/Explosion        1          1  ...   

                    targtype1_mapped  natlty1  property INT_MISC  \
4955                       Utilities      4.0         1        0   
4956         Educational Institution    138.0         0        0   
4957  Religious Figures/Institutions    147.0         0        0   
4958                         Unknown     45.0         0        0   
4959                        Business    138.0         1        0   

      targsubtype1  weaptype1  specificity  attacktype1  nkill  weapsubtype1  
4955         107.0          6            1            3    0.0          16.0  
4956          50.0         13            1            6    1.0          11.0  
4957          85.0          5            4            6    0.0           5.0  
4958          35.0          6            4            3    0.0          16.0  
4959          11.0          6            2            3    0.0          16.0  

[5 rows x 23 columns]

#### Data Visualization and Exploratory Data Analysis

Explain what the chart shows and what insights can be seen from it

7.1) It is required for you to use statistical techniques, univariate, bivariate and multivariate analysis, etc. You may use techniques that are not discussed in synchronous lectures or in class.
7.2) It is NOT required for you to implement Machine Learning techniques. But you are encouraged to try to implement.

#### Summary and Recommendations

- summarize findings

#### Future Work

9.1) Include a Gantt chart that has a timeline of the next steps if you are given 2 to 4 months to continue this project/research.
9.2) If you did not implement machine learning technique/s, what are the candidate machine learning technique/s that you plan. Give an explanation of what are the possible benefits or improvements of these machine learning technique/s from what you have implemented in step 7 (EDA and Statistics) and details (e.g. hyperparameter tuning, different settings, and experiments).
9.3) If you were able to implement machine learning technique/s, what are other or more advanced techniques that you plan to implement? Include details (e.g. hyperparameter tuning, different settings, and experiments) and how would these other or more advanced techniques compare to the current machine learning techniques that you were able to implement.

#### References

You are encouraged to look at existing solutions online and learn from them (please cite). You also are encouraged to explore other techniques that were not discussed in class and cite all of your references.

START (National Consortium for the Study of Terrorism and Responses to Terrorism). (2022). Global Terrorism Database 1970 - 2020 [data file]. https://www.start.umd.edu/gtd